### DISCUSSION TITLE
Precompilation and more

### DISCUSSION THREAD

**Discussion by jbcaillau** [control-toolbox/control-toolbox.github.io#51] [Ideas]
@ocots @PierreMartinon @joseph-gergaud As of today, we haven't taken into account precompilation (in the broad sense) issues. As a matter of fact, `using OptimalControl` is quite (very) long. Below some (very) interesting links pointed by @amontoison 🙏🏽🙏🏽🙏🏽 on the topic (profiling / improving performance when `add(ing)` (in the package manager) and `using` (within the package manager):
- [PrecompileTools](https://github.com/JuliaLang/PrecompileTools.jl)
- [JET](https://github.com/aviatesk/JET.jl)
- [example with JET on Krylov.jl](https://github.com/JuliaSmoothOptimizers/Krylov.jl/discussions/698)
- [SnoopCompile](https://timholy.github.io/SnoopCompile.jl/stable/)
- [JuMP precompilation script](https://github.com/amontoison/JuMP.jl/blob/master/src/precompile.jl) [^1]
- [see also](https://github.com/amontoison/JuMP.jl/blob/master/src/JuMP.jl#L991-L1067) (uses `SnoopCompile`)
- [must read on Julia 1.9](https://julialang.org/blog/2023/04/julia-1.9-highlights/)

[^1]: such scripts can be run ahead of `using` the package, preparing as much stuff as possible; of course, static types can be taken care of, not dynamic (dispatch...) ones

**Comment by jbcaillau** [comment]
See also (merci @ocots) : https://programmation.developpez.com/actu/354592/Annonce-de-Julia-v1-10-la-nouvelle-version-du-langage-de-programmation-apporte-des-ameliorations-en-termes-de-rapidite-et-de-confort-pour-les-developpeurs/

**Reply by PierreMartinon** [reply]
Just tested CTDirect (not the whole OptimalControl) on 1.10.1 (linux mint)
I got a precompilation error from CTBase, however tests seem to run fine and almost twice faster than on 1.9 (45-50s vs 75-80s).

> WARNING: Method definition apply_recipe(Base.AbstractDict{Symbol, Any}, CTBase.OptimalControlSolution, Union{Symbol, Tuple{Symbol, Integer}}, Union{Symbol, Tuple{Symbol, Integer}}) in module CTBase at /home/pierre/.julia/packages/RecipesBase/BRe07/src/RecipesBase.jl:296 overwritten on the same line (check for duplicate calls to `include`).
ERROR: Method overwriting is not permitted during Module precompilation. Use `__precompile__(false)` to opt-out of precompilation.
